types/coin.go

```diff
@@ -508,15 +513,49 @@ func ParseCoin(coinStr string) (coin Coin, err error) {
 
 	denomStr, amountStr := matches[2], matches[1]
 
-	amount, ok := NewIntFromString(amountStr)
-	if !ok {
-		return Coin{}, fmt.Errorf("failed to parse coin amount: %s", amountStr)
-	}
-
 	if denomStr != strings.ToLower(denomStr) {
 		return Coin{}, fmt.Errorf("denom cannot contain upper case characters: %s", denomStr)
 	}
 
+	// test if the amount is specified using scientific notation
+	scinMatches := reScinAmt.FindStringSubmatch(amountStr)
+
+	ok := false
+	var amount Int;
+
+	if len(scinMatches) == 4 {
+
+		// Parse scientific notation such as 1.23e+12stake
+		var f float64
+		var dec int
+		
+		if f, err = strconv.ParseFloat(scinMatches[1], 64); err != nil {
+			return
+		}
+		
+		if dec, err = strconv.Atoi(scinMatches[3]); err != nil {
+			return 
+		}
+
+		// adjust dec by the length of the decimal part in the number
+		d := reDecPart.FindStringSubmatch(scinMatches[1])
+		if len(d) > 1 {
+			l := len(d[1])
+			if l > dec {
+				return Coin{}, fmt.Errorf("amount %s has too many digits after the decimal point resulting in precision loss",amountStr)
+			}			
+			f, dec = f * math.Pow10(l), dec - l
+		}
+		amount, ok = NewIntWithDecimal(int64(f), dec), true
+
+	} else {
+		amount, ok = NewIntFromString(amountStr)
+	}
+
+	if(!ok) {
+		return
+	}
```

Comment:
Errors should be returned immediately - the `ok` variable should only be defined and used within the `if` statement (aka line 523 shouldn't need to exist, nor this return statement)

Reply:
following this convention now (also this is my first code in golang, please bear with me)

Reply:
hehe np! This is actually a new convention we've adopted in the code base - it's not technically incorrect to do what you've done